connect/api/src/main/java/org/apache/kafka/connect/data/ConnectSchema.java

```diff
@@ -118,6 +120,7 @@ public ConnectSchema(Type type, boolean optional, Object defaultValue, String na
 
         this.keySchema = keySchema;
         this.valueSchema = valueSchema;
+        this.hash = Objects.hash(type, optional, defaultValue, fields, keySchema, valueSchema, name, version, doc, parameters);
```

Comment:
Do we actually want this to always be precomputed or just on demand (i.e. store as an `Integer` that's `null` until the first time `hashCode` is called)?

Reply:
Avro's `Schema` computes and caches the hash code [on demand](https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/Schema.java#L354-L358). Doing that does add overhead (non-synchronized computing and setting of the hash code to a non-volatile field might mean it's computed a few times), so considering that Connect will often need the hash code is it really worth optimizing for that case? Certainly the Avro converter will use the `hashCode()` method for every `ConnectSchema` it sees.

Reply:
+1 on precomputing hash code.

Reply:
I asked because Connect doesn't make any guarantees about thread safety of that class in the first place (within the framework, everything is sufficiently locked, though people may be implicitly making assumptions in connectors since it was intended to be immutable anyway). We're making micro-optimizations here without benchmarks, so we really do need to take care.

I can think of cases we're definitely increasing overhead here (e.g. any transformation that filters out data would probably mean schema hash is never computed whereas now it is). I can't think of any cases in the connectors we've built where you would end up double computing (they pretty much all do any sort of schema hashing in a single thread because the whole point of tasks in connect is to handle parallelism at the fmwk level automatically for you, and any threading usually is internal to the library handling the connection to the other data storage system; the caching in converters would also generally be safe except for static schemas, and in that case the overhead is not worth worrying about and I think truly static schemas in connectors are pretty rare anyway).

Reply:
If I use Integer (initially null) to represent the hash code, each subsequent hashCode() would involve a check (against null) and unboxing.
Just want to confirm that it is Okay going this route.